src/hash/blake/mod.rs

```diff
@@ -0,0 +1,261 @@
+use super::{ElementHasher, HashFn};
+use crate::{
+    ByteReader, ByteWriter, Deserializable, DeserializationError, Digest, Felt, FieldElement,
+    Serializable,
+};
+use core::{
+    mem::{size_of, transmute},
+    ops::Deref,
+    slice::from_raw_parts,
+};
+
+#[cfg(test)]
+mod tests;
+
+// BLAKE3 256-BIT OUTPUT
+// ================================================================================================
+
+/// 256-bit output of a blake3 function.
+#[derive(Debug, Default, Copy, Clone, Eq, PartialEq)]
+pub struct Blake3Digest256([u8; Self::SIZE]);
```

Comment:
Why create 3 different digests rather than one single digest which could look something like this:
```Rust
pub struct Blake3Digest<const N: usize>([u8; N]);
```
And then just use the appropriate `N` for each case?

Reply:
We have an exception for the 256 bits output. The `Digest::as_bytes` instruction will just copy its inner bytes to return a `[u8; 32]`, while the others will have to perform an expansion (24 -> 32, etc)

I think if I just code the `expand_bytes` with a branching with just constants as arguments, the compiler will end up not branching. Testing that now!

Reply:
The complete solution is to update also the digest trait in winterfell to take associated constants (that are now supported)

https://github.com/novifinancial/winterfell/blob/d0771ef44cc59a8f1b4bb9d27f552dda1d70a615/crypto/src/hash/mod.rs#L63-L74

Something like

```rust
pub trait Digest<const N: usize>:
    Debug + Default + Copy + Clone + Eq + PartialEq + Send + Sync + Serializable + Deserializable
{
    /// Returns this digest serialized into an array of bytes.
    fn as_bytes(&self) -> [u8; N];
}
```

Do you think it is desirable to tackle this in this issue? We can go one of the two ways: create our own `Digest` in this lib and start to use it, or start a change request in winterfell

Reply:
Let's create a PR in Winterfell to update this. We won't be able use this until Winterfell v0.5 is released (which will likely take time) - but at least later on we'll do it in the preferred way.

For now, what you have here works very well, I think.